Feature/controller add turris omnia rdkb support

morantr opened the pull request:
Port controller changes from turris omnia support branch (#307).

tomereli reviewed (changes requested):
First of all, please add a PR description which summarizes the changes

arnout commented on controller/CMakeLists.txt:
```diff
@@ -63,16 +66,21 @@ message(STATUS "CMAKE_PREFIX_PATH - ${CMAKE_PREFIX_PATH}")
-set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -O2 -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Wl,-S -fPIE -fPIC")
+
+# Default Compiler flags
+set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wall -Wextra -Werror -pthread -Wno-comment  -Wno-unused-parameter -Wno-shift-negative-value -Wno-missing-field-initializers -fPIC")
```
Instead of passing `-Wno-shift-negative-value`, wouldn't it be better to fix the issue? Should be easy enough to fix.

The other exclusions are OK, they're also used in other places.

morantr replied:
I think @rmelotte did it. @rmelotte do you know how to fix it?

rmelotte replied:
I don't remember where the issue was, let me check.

arnout commented on controller/CMakeLists.txt:
```diff
@@ -63,16 +66,21 @@ message(STATUS "CMAKE_PREFIX_PATH - ${CMAKE_PREFIX_PATH}")
 endif()
+
+# Release flags
+set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -s -O2 -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Wl,-S -fPIE")
```
I don't think there's any reason to move CMAKE_C_FLAGS_RELEASE after the compiler-specific flags, it should work either way.

However, I do thing CMAKE_CXX_FLAGS should move, because now it doesn't include the -Wno-address-of-packed-member.

morantr replied:
switched between lines 72 and 83

arnout commented on controller/src/beerocks/cli/beerocks_cli_main.cpp:
```diff
@@ -53,14 +53,14 @@ static void set_cursor(int n, int m)
 }
 
-static void sigterm_handler()
+static void sigterm_handler(int signum)
```
This change (all changes to this file) should be in a separate commit, with a commit message that explains why.

morantr replied:
I asked @vitalybu about this, and he said that it didn't compile without it on Turris without it.
That is the reason it is not in a separate commit.

arnout commented:
Commit d0b9cea797552e3a34cee229f713bf1ae907ad84 was originally authored by Raphaël (commit 6dc89b8ebd810c9893787d633ecfa0b62a1d2adb). You can't claim you authored it, you must preserve Raphaël's authorship. If you make minor changes to it, you should add your own Signed-off-by but keep the original author. If you make major changes, you can take authorship, but then Raphaël's Signed-off-by should still be there.

To fix this, you can either restore the author with `git commit --amend --author='Raphaël Mélotte <raphael.melotte@essensium.com>'`, or you can recover the original metadata with `git commit --amend -s -c 6dc89b8ebd810c9893787d633ecfa0b62a1d2adb`.

arnout commented:
> you must preserve Raphaël's authorship.

There are others as well, e.g. 26f3b5cebe29ef4c06ace57f2c5932d6a1e03844 was authored by @vitalybu. Please review all commits and restore authorship.

arnout commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1503,7 +1503,7 @@ int cli_bml::analyzer_init(std::string remote_pc_ip)
             }
         }
-        if (t < 19) {
+        if (t < 1) {
```
I believe this change is no longer needed because it has been fixed in the analyzer. @rmelotte can you confirm?

rmelotte replied:
It's not needed anymore with the "controller: remove "unknown IP" states" commit (currently d0b9cea).
It was only needed because the controller didn't send updates to the analyzer when an event occurred (a client connected, disconnected,...).

tomereli replied:
@morantr please remove this commit in that case

morantr replied:
done

arnout commented on controller/src/beerocks/master/tasks/channel_selection_task.cpp:
```diff
@@ -844,11 +845,12 @@ void channel_selection_task::work()
+            database.set_hostap_cac_completed(hostap_mac, true);
+        } else if (wireless_utils::is_dfs_channel(hostap_params.channel)) {
                 wait_for_cac_completed(hostap_params.channel, hostap_params.bandwidth);
             }
```
There's an indentation error introduced here (in commit "controller: Add support for passive mode") that gets fixed in a later commit.

morantr replied:
will do clang-format on that commit

arnout commented on controller/src/beerocks/master/tasks/optimal_path_task.cpp:
```diff
@@ -329,8 +337,8 @@ void optimal_path_task::work()
 
-        if (responsiveness_precentage_11k != 100) {
-            TASK_LOG(DEBUG) << "11k measurement request responsiveness is less than 100%";
+        if (responsiveness_precentage_11k < 50) {
```
This is a good example of a change that should be done in a completely separate commit, with a good commit message that explains why 50% is "better" than 100%. I guess the reasoning is something like: "Some STAs are able to perform 11k beacon requests, but not for all cases, e.g. not in a different band. In this case, we still want to use the 11k responses we do get."

tomereli replied:
I would go even further and add this explanation here in a comment. Makes reading the code much easier.

morantr replied:
I wish I had a good explanation for it, but there isn't one. This is an optimization that @kantera800 asked me to do since most of the time we didn't get 100%, and optimal_path stopped, so changing to 50% made the demo work better, but it isn't necessarily "better" because the reason we don't have 100% might be because the station "didn't want" to reply. So our decision might be wrong.
So I wouldn't write it in the commit message nor in a comment (stupid thing to write IMO).
But if you have and a specific comment you want me to add as a comment or to the commit message, please write it down and I'll add it.

Either way, all of the changes in this commit are pretty small and it has been done a part of optimal_path optimization so I preferred doing it all in a single commit. I don't think that separating it will give extra benefit, and there is no point in doing it now.

tomereli replied:
Bottom line - its a HACK we don't know why is working and might not be working at all?
So at least add a comment (and possibly a task) which explains that this is done as part of demo and should be revised in the future.

morantr replied:
added a comment to the code.

arnout commented on controller/src/beerocks/master/tasks/optimal_path_task.cpp:
```diff
@@ -78,6 +78,18 @@ void optimal_path_task::work()
+        if (steer_vaps.find(current_vap_name) == std::string::npos) {
+            TASK_LOG(INFO) << "current hostap vap:" << current_vap_name << " mac:" << sta_mac
+                           << "is not in steer list:" << steer_vaps << " aborting optimal task"
+                           << std::endl;
```
I haven't tested this, but this looks like optimal path task will be disabled entirely except for the VAP+sta_mac combinationas that are explicitly mentioned in the controller config file. I don't see how that can be a useful feature...

tomereli replied:
This feature came as a request from a customer (a big one). Perhaps the reasoning is that for some clients / VAP combinations, they might have different algorithms than the optimal path to do the steering?

morantr replied:
@tomereli is that a question for me? Because I don't have a good answer.

tomereli replied:
It was an answer for @arnout. I think @simonbaren is better equipped to answer that.

arnout commented on controller/src/beerocks/master/beerocks_master_mrouter.cpp:
```diff
@@ -118,7 +118,12 @@ bool master_mrouter::handle_cmdu(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx)
     std::shared_ptr<sMasterSockets> soc = nullptr;
     auto uds_header                     = message_com::get_uds_header(cmdu_rx);
-    uint16_t length                     = uds_header->length;
+    if (uds_header == nullptr) {
```
More a coding style question than an actual review comment:

Somebody (@tomereli ?) said at some point in reviews that for smartptrs, you should use `if (!ptr)` rather than `if (ptr == nullptr)`. If that is the case, I think we should do the same for normal pointers.

Whoever made that comment should update CODINGSTYLE.md with that information.

tomereli replied:
@morantr I think you said that KW fails on such checks, and we need to use `if (!ptr)` rather than `if (ptr == nullptr)`.

morantr replied:
Yes, I said it. But it only applies to smart pointers, because I remember tons of complaints about it when fixing Klocwork issues. For some reason, Klocwork didn't like.
Anyway, it does not apply to regular pointers which is the case here.

arnout replied:
> Anyway, it does not apply to regular pointers which is the case here.

That's why my original comment was: "I think we should do the same for normal pointers." Unless, of course, klockwork doens't like `!ptr` for normal pointers.

arnout commented on controller/src/beerocks/master/db/network_map.cpp:
```diff
@@ -497,10 +497,20 @@ void network_map::send_bml_bss_tm_req_message_to_listeners(db &database,
+    event->type = BML_EVENT_TYPE_BSS_TM_REQ;
+    auto size   = sizeof(BML_EVENT);
+    event->data = GET_MESSAGE_POINTER(BML_EVENT_BSS_TM_REQ, response->buffer(0), size);
+    if (event->data == nullptr) {
```
GET_MESSAGE_POINTER cannot yield nullptr because it adds size to the pointer. Well, it can be nullptr if the pointer happens to be minus maxint, but that is unlikely...

In other words, I don't think this check is needed.

Same in a few other places.

morantr replied:
removed

arnout commented on controller/src/beerocks/master/db/network_map.cpp:
```diff
@@ -497,10 +497,20 @@ void network_map::send_bml_bss_tm_req_message_to_listeners(db &database,
-    auto size             = sizeof(BML_EVENT);
-    event->data           = GET_MESSAGE_POINTER(BML_EVENT_BSS_TM_REQ, response->buffer(0), size);
+    auto event = (BML_EVENT *)response->buffer(0);
+    if (event == nullptr) {
```
I don't know how relevant this check is, because there is no way that event can be nullptr. It can only happen if init() wasn't called (or the argument to buffer() is larger than the allocated size of buffer, but since it's 0 that can't happen).

But OK, no harm I guess.

tomereli replied:
And here as well - please change to `if (!event)`

morantr replied:
But it is not a smart pointer, so it is not relevant.
Now that I look at it, I notice that the cast needs to be changed to a reinterpret_cast.

tomereli replied:
should have been a (nitpick)

arnout reviewed (approved):
As usual, I have a lot of feedback, but nothing critical, so I'm approving. Please do the following:

- rebase on master;
- remove the analyzer speedup and load_steer_on_vaps patches;
- correct authorship.

If the removed patches are needed after all, we can still recover them from the dev/turris-omnia branch.

tomereli commented on controller/config/beerocks_controller.conf.in:
```diff
@@ -30,7 +30,7 @@ load_ire_roaming=1
 
 #   Measurement feature:  
-load_diagnostics_measurements=0
+load_diagnostics_measurements=1
```
Can you explain in the commit message what are the diagnostics measurements and why they are enabled now?

morantr replied:
I don't know why they enabled it. If I remember correctly you disabled it since it flooded the controller logs with timeout prints.
This task is responsible for periodically ask for statistics on hostapd and it's connected stations.
My guess is that the reason is to see `rx_rssi` value being updated in the connection map.
@vitalybu ?

tomereli replied:
Please talk to Vitaly and add a comment so we can resolve this conversation.

morantr replied:
Ok, @vitalybu confirmed to me that he did it to update the rx_rssi. I'll add it to the commit message.

morantr replied:
updated the commit message.

tomereli commented on controller/src/beerocks/bml/bml_defs.h:
```diff
@@ -68,8 +68,7 @@ extern "C" {
 #define BML_NODE_STATE_DISCONNECTED 0
```
Wasn't this commit by @rmelotte ?

morantr replied:
re-authored

tomereli commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2202,6 +2202,38 @@ bool db::get_hostap_is_acs_enabled(std::string mac)
     return n->hostap->is_acs_enabled;
 }
 
+bool db::set_hostap_passive_mode_enabled(std::string mac, bool enable)
```
The whole of this commit should be removed. The only mode we support from now on is passive mode, so the controller doen't need to be notified.

morantr replied:
For now, it is needed because some parts of the code are using it.
I will remove it nicely in a different PR right after I'll finish with the merging the Turris Omnia changes.

tomereli replied:
OK.

tomereli commented on controller/src/beerocks/master/tasks/task.cpp:
```diff
@@ -33,11 +33,14 @@ void task::response_received(std::string mac, beerocks_message::eActionOp_CONTRO
     }
+
+    // Handle the response even if we are not expecting it
+    handle_response(mac, action_op, cmdu_rx);
```
This looks like a real hack and I'm not sure it's worth it. Why did we do this change in the demo? At least we should add a print when such a thing happens, but I rather drop this commit altogether.

morantr replied:
It is an optimization for the optimal path task.
Imagine the optimal path task starting to ask a measurement on each BSSID at the time.
Now, let's assume that the task requested measurement on BSSID_A and waited for a response, but reached timeout, and then started to ask for measurement on BSSID_B, but while doing so, received the response for BSSID_A. If it is good, it will be used, and will save an additional measurement to BSSID_A.
If you take into account that each measurement waits for a response to at most 6 seconds and that the second measurement might be bad (when you already received a good measurement), you have a lot to lose here.

tomereli replied:
still feels like a hack, but OK I guess..

tomereli commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2929,6 +2929,11 @@ bool db::set_node_channel_bw(const std::string &mac, int channel, beerocks::eWiF
         }
     }
 
+    LOG(INFO) << "set node " << mac << " to channel=" << channel << ", bw=" << int(bw)
```
How often does this happen? We don't want the log to become too clattered.
If you decide to keep this commit, I suggest explaining the scenario in which you felt this print would help so anyone reading this can understand why it is actually needed (me included).

morantr replied:
It should happen once when receiving a slave join message which initiates the channel selection.
It also happens when sta is connected, and we update it's channel in the node, but IMO we need to remove the channel from sta node since it will always have the same channel as its parent radio which makes it redundant.

Please try to understand that I didn't write most of the code here, so I don't always have an answer or explanation of why some change has been done, I can only guess.

tomereli reviewed (approved):
I'm approving so not to block this, since too many development tasks rely on this. Also @arnout already approved so I guess its fine to merge :)

morantr committed:
controller: Adjust the controller to support Turris Omnia platform

cmake adjustments to compile and run successfully prplMesh on Turris
Omnia platform.

Additional changes:
- Ignore warnings "left shift of negative value
[-Werror=shift-negative-value]" which appears when using GCC7.
- Explicitly link with MapfTransport.
When building for OpenWrt, for some reason it doesn't find MapfTransport
which is a transitive dependency from btl. So add it explicitly to the
link of agent and controller for the time being.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
controller: Enable diagnostics measurements on configuration file

Enabling diagnostics measurements enables the statistics polling task
which responsible to poll periodically from the Agent statistics on
the radio and its connected station and saves them on the controller
database.
The main reason for doing it is to update the rx_rssi field which
is visible on the CLI conn_map.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

rmelotte committed:
controller: remove "unknown IP" states

The Turris Omnia doesn't have the IP addresses of the clients, which
make the optimal_path_task fail (it checks for a CONNECTED state, but
the state is CONNECTED_IP_UNKNOWN).

There shouldn't be a difference between a connected state and a
connected state without a known IP, as prplMesh is supposed to work
just as well without IPs.

In cases where it's really needed to know if the device has an IP or
not, it's still posssible to check if the IP has the "0.0.0.0" value.

Note that this commit may need additional fixes, as there are places
where "no IP address" is represented by an empty string (like in the
database), and other places where "no IP address" is represented by
the "0.0.0.0" IP.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
controller: Add support for "load_steer_on_vaps" feature

This feature enables to optimal path task to steer clients only between
preconfiged bssids.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
controller: Add missing nullptr checks

Add missing nullptr checks on various places on the code.
Also, fix functions wrong return statements.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
controller: Move band steering out of rdkb bml code scope

Move band steering out of rdkb bml code scope and thus,
support controlling client_band_steering on non-RDKB platforms.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
controller: Add support for passive mode

Add support for passive mode, which allow to run the controller without
activly change anything on the agent side (like channel for example).

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
controller: cli: Fix bug caused to wrong eth mac presentation

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
controller: bml: Replace less-standard type `uint` with `uint8_t`

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
controller: Allow tasks to receive responses when not waiting for it

Optimization.
Allow tasks to still get a response even if the task is no longer
pending for it. This is might be helpful when a task sends multiple
synchronized requests and, and we don't want to set a very long timeout,
just to make sure we won't miss the response.
This allows to set lower timeouts, and still get the response if the
the task didn't start to process all of the responses.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
controller: Fix a wrong presentation of channel and bw

Fix bug that caused to wrong bw value in the controller database.
Also, add a print of the channel.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
controller: Setting agent socket

Setting the agent socket will always take place regardless if it is
a backhaul manager son_slave or not.
Setting it on specific son_slave was relevant when the socket was
associated with the radio mac, and not bridge mac like now.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
controller: Fix bug of copy mac address from the wrong place on struct

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
controller: client steering: use 0 as messageId

The task ID was previously used as the messageId. The consequence was
that the ieee1905_transport was detecting the steering request as
duplicate, because it used the same MAC src, and the same messageId.

This commit uses 0 as messageId, so that it's generated automatically,
and that the packets are no longer flagged as duplicate.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
controller: optimal_path tweaks

- Handle the response to a task even if we are not expecting it.
- Stop waiting to 11k response if an invalid response has been received.
- Change responsiveness percentage of 11k measurements to be greater
than 50% instead of 100% to steer station based on 11k measurements.
- Modify/Add/Remove debug prints.
- Change ranged for loops to use const reference.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr committed:
controller: Add debug print on set_node_channel_bw()

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

morantr commented:
> Commit [d0b9cea](https://github.com/prplfoundation/prplMesh/commit/d0b9cea797552e3a34cee229f713bf1ae907ad84) was originally authored by Raphaël (commit [6dc89b8](https://github.com/prplfoundation/prplMesh/commit/6dc89b8ebd810c9893787d633ecfa0b62a1d2adb)). You can't claim you authored it, you must preserve Raphaël's authorship. If you make minor changes to it, you should add your own Signed-off-by but keep the original author. If you make major changes, you can take authorship, but then Raphaël's Signed-off-by should still be there.
> 
> To fix this, you can either restore the author with `git commit --amend --author='Raphaël Mélotte <raphael.melotte@essensium.com>'`, or you can recover the original metadata with `git commit --amend -s -c 6dc89b8ebd810c9893787d633ecfa0b62a1d2adb`.

I didn't know I can change authorship on commits I am pushing. Thank you for explaining to me how to do it.
Please try to understand that in order to make this PR I had to use some folder comparing software  (like beyond compare), on the original Turris branch which has 64 commits, and take only the relevant code by moving it from left to right, and then separate nicely to commit which is very hard thing to do, and that I didn't try to claim changes that are not rightfully mine.
I changed the commit authorship on that specific commit you asked me, but not on other commits.
What I wanted to ask is that if some of you see a commit that you think you are the rightful owner of it, please help me and change the authorship.
@vitalybu, @arnout, @rmelotte

morantr commented:
@rmelotte Do you have an idea why test_flows.sh is failing here? on my machine, it all passes.

tomereli commented:
> @rmelotte Do you have an idea why test_flows.sh is failing here? on my machine, it all passes.

It's not a real failure since Branch and Pull Request passed. Can we merge this or do we wait for anything else (like commits ownership changes)?

rmelotte commented:
> It's not a real failure since Branch and Pull Request passed. Can we merge this or do we wait for anything else (like commits ownership changes)?

Is there a timing issue with the CI tests? Can we avoid it or should we just increase the wait time?
